### ISSUE TITLE
cmd/go: typos are slow

### ISSUE THREAD

**Issue by adonovan** [golang/go#26874]
The new semantics of `go build` mean that it takes 12 seconds to report a typo in a package name.

```
$ pwd
/home/adonovan/got/src/golang.org/x/tools

$ time go build ./cmd/ssadump
real	0m0.185s
user	0m0.332s
sys	0m0.140s

$ time go build ./cmd/ssadum
go: finding golang.org/x/tools/cmd/ssadum latest
go: finding golang.org/x/tools/cmd latest
go: import "golang.org/x/tools/cmd/ssadum": cannot find module providing package golang.org/x/tools/cmd/ssadum

real	0m12.422s
user	0m0.200s
sys	0m0.096s
```

**Comment by oryanmoshe** [comment]
What has changed that caused it?
Was it shorter before?

**Comment by cespare** [comment]
@oryanmoshe yes, module semantics have made this slow. On my machine (Go 1.10) it takes about 10ms if I misspell the name.

**Comment by oryanmoshe** [comment]
@cespare
Thanks, I didn’t know what to look for until you said modules.. Didn't get the chance to dive into modules internals yet. 

I'll look at it and see if I can spot why it's taking so long.

**Comment by bcmills** [comment]
That's especially odd, since according to my understanding, patterns starting with `./` should only match packages in the current module (see discussion on #26317).

**Comment by bcmills** [comment]
@alandonovan, are you seeing this slowness for all builds, or only those with modules enabled?

**Comment by rsc** [comment]
The reason is clear from the output: it tries to do a lookup for a different module that might provide ssadum. We need to avoid that.
